aot_autograd: keep input mutations on requires_grad=True tensor out of the graph for inference

bdhirsh opened the pull request:
The original behavior of torch.compile w.r.t. input mutations maintains that if an input to a graph was mutated, **and** requires grad, we will keep the input mutation outside of the graph and replay it at runtime.

This is important because, e.g., an input can have outstanding aliases, and mutating the input in eager mode will cause autograd to change the `grad_fn` of all outstanding aliases.

It looks like landing https://github.com/pytorch/pytorch/pull/111347 changed this behavior slightly:
* The linked PR makes it possible for AOTAutograd to go down the inference code path, even if some inputs require grad (because all of the outputs of the graph were seen to not require grad)
* AOTAutograd's logic in the inference code path today is to **always** keep input mutations in the graph.

This PR fixes that regression: regardless of inference vs. training, we should always keep input mutations outside of the graph if the input requires_grad.

cc @tugsbayasgalan (for review)

cc @oulgen (related to some of your q's around triton kernel handling)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113417
* #113277
* __->__ #113584
* #113416
* #113267

bdhirsh committed:
aot_autograd: keep input mutations on requires_grad=True tensor out of the graph for inference

[ghstack-poisoned]

tugsbayasgalan commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3192,36 +3197,16 @@ def runtime_wrapper(*args):
-                assert (
-                    runtime_metadata.num_mutated_inp_runtime_indices == num_mutated_runtime_inps
-                )
+        if keep_input_mutations and trace_joint:
```
Nice!!

tugsbayasgalan commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1463,7 +1463,14 @@ def view_avoid_dupes_with_primals(t):
+                requires_grad=info.requires_grad
+                # MUTATED_OUT_GRAPH corresponds to any input mutations that happen outside the graph.
+                # this can also include metadata mutations, and inputs that do not require grad,
+            ) == MutationType.MUTATED_OUT_GRAPH and info.mutates_data and info.requires_grad
```
What would happen for inputs that are marked with MutationType.MUTATED_OUT_GRAPH but not fall under "info.mutates_data and info.requires_grad"

bdhirsh replied:
An example would be an input that had a metadata mutation - we always keep metadata mutations outside of the graph, regardless of whether they require grad.

In that case, we still don't want to create a tangent (backward input) for that input (we only need to create tangents for inputs that require grad and have data mutations).

I think the main reason I added this extra check is because @oulgen is doing some work to add mutations that can be "hidden" from autograd, that we want to keep in the graph. And for that case, we also don't want to create tangents for those updated inputs in the backward.

bdhirsh committed:
Update on "aot_autograd: keep input mutations on requires_grad=True tensor out of the graph for inference"

The original behavior of torch.compile w.r.t. input mutations maintains that if an input to a graph was mutated, **and** requires grad, we will keep the input mutation outside of the graph and replay it at runtime.

This is important because, e.g., an input can have outstanding aliases, and mutating the input in eager mode will cause autograd to change the `grad_fn` of all outstanding aliases.

It looks like landing https://github.com/pytorch/pytorch/pull/111347 changed this behavior slightly:
* The linked PR makes it possible for AOTAutograd to go down the inference code path, even if some inputs require grad (because all of the outputs of the graph were seen to not require grad)
* AOTAutograd's logic in the inference code path today is to **always** keep input mutations in the graph.

This PR fixes that regression: regardless of inference vs. training, we should always keep input mutations outside of the graph if the input requires_grad.

cc tugsbayasgalan (for review)

cc oulgen (related to some of your q's around triton kernel handling)




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
